dill/session.py

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/env python
```

Comment:
### Size filter

Discussion history:

> **mmckerns:** Another interesting filter might be very large objects...
> 
> **leogama:** This is an even harder problem. `sys.getsizeof()` gives reasonable values for some objects, like an array, but not others, like a list of arrays or a complex object from `scikit-learn`. What I can think of is 1) to "`dumps`" the object to a bytes string; 2) check the object's serialized size; 3) discard it or write the bytes directly to the stream, taking care of the module's dictionary opcodes manually as in the unpickable case or wrapping it in a lightweight dill class that would "`loads`" it at unpickling.
> 
> **mmckerns:** Arrays work here because they are containers of ctypes. Lists and other python containers hold references, so the lists themselves are small, but some of the referenced items are not. I think it would be fair to: (1) `getsizeof` an object, and replace anything that registers as bigger than some cutoff with a stub, and/or (2) pickle an object to a byte string as you suggest, then make a decision based on the resultant size. The latter might be a slow process... where filtering by size, especially in the latter case, is going to be inefficient. You already do something similar to `getsizeof` in the trace, I can imagine that it could be leveraged in a filter.

Reply:
So we have 3 possible strategies by now:
- The naive `getsizeof` approach, that could be incremented to look inside container objects (ever recursively, if we take care of recursive structures). This should be reasonably fast.
- The `serialize → check size → write` approach that I mentioned. This would be the slowest one, but maybe not so much as the serialization would be computed just once per object (and maybe it could be aborted midway for large objects).
- The `serialize → write → check size → roll back if too big` approach, inspired by what `logger.trace()` does. This one may or may not be efficient depending on the size of the large objects being discarded. I'm not so sure the serialization of such objects could be aborted as in the previous case. The stream would need to have a `seek()` method and framing would likely have to be turned off.

Only the first approach can be implemented as a simple filter. I would create a filter factory for it, like `ipython_filter()`:

```python
def size_filter(size):
    import re, sys
    match = re.fullmatch(r'(\d+)\s*(B|[KMGT]i?B?)', size, re.IGNORECASE)
    if not match:
        raise ValueError("invalid 'size' value: %r" % size)
    coeff, unit = match.groups()
    coeff, unit = int(coeff), unit.lower()
    if unit == 'b':
        size = coeff
    else:
        base = 1024 if unit[1:2] == 'i' else 1000
        exponent = 'kmgt'.index(unit[0]) + 1
        size = coeff * base**exponent
    def exclude_large(obj):
        return sys.getziseof(obj) > size
    return exclude_large
```